expmod/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+extern crate ewasm_api;
+extern crate num_bigint;
+
+use num::{One, Zero};
+use num_bigint::{BigInt, Sign};
+use std::cmp;
+use std::ops::Shl;
+
+fn f(x: usize) -> u64 {
+    let cost = match x {
+        0...64 => x * x,
+        65...1024 => (x * x / 4) + 96 * x - 3072,
+        _ => (x * x / 16) + 480 * x - 199680,
+    };
+    cost as u64
+}
+
+const G_QUAD_DIVISOR: u64 = 20;
+
+fn calculate_cost(lm: usize, lb: usize, exp: &BigInt) -> u64 {
+    let le = exp.to_bytes_be().1.len();
+    let lep: u64 = match le {
+        0...32 => {
+            if exp.is_zero() {
+                0u64
+            } else {
+                // Number of bits == ⌊log2()⌋ + 1
+                exp.bits() as u64 - 1
+            }
+        }
+        _ => {
+            let low32bytes = exp.modpow(&BigInt::one(), &BigInt::one().shl(256));
+            let loglow32bytes = low32bytes.bits() as i64 - 1;
+            if low32bytes > BigInt::zero() {
+                8 * (le as u64 - 32) + loglow32bytes as u64
+            } else {
+                8 * (le as u64 - 32)
+            }
+        }
+    };
+
+    (f(cmp::max(lm, lb)) * cmp::max(lep, 1 as u64)) / G_QUAD_DIVISOR
+}
+
+fn expmod(base: &BigInt, exp: &BigInt, modulus: &BigInt) -> Vec<u8> {
+    let x = if modulus.is_zero() {
+        BigInt::zero()
+    } else {
+        base.modpow(exp, modulus)
+    };
+    let (_, mut data) = x.to_bytes_be();
+
+    /* padded on 32 bytes */
+    let mut padded = vec![0u8; (32 - (data.len() % 32)) % 32];
+    padded.append(&mut data);
+    padded
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn main() {
```

Comment:
I think the main part of testing is about decoding the inputs, which is done in this function.

I'd prefer to split this into function which can be tested using unit tests.

Reply:
most of it depends on the output of calldata_copy and I suspect the BigInt library to be very inefficient so I don't want to pass those BigInts back and make things worse. Although it hasn't been submitted yet, I have tests for these in the geth PR.

Reply:
This is a trick some other precompiles use ([for example](https://github.com/ewasm/ethereum-bn128.rs/blob/master/src/lib.rs)
```
let mut padded_input = input.chain(io::repeat(0));
```